### ISSUE TITLE
Scoped dependency gets same instance every request with GetService<T>();

### ISSUE THREAD

**Issue by Menighin** [dotnet/efcore#14840]
I'm trying to achieve the the scenario described here: https://stackoverflow.com/questions/54888399/getting-user-information-in-a-middleware-and-injecting-it

To do so, I'm trying to implement things like what I imagine to be `HttpContextAccessor`.
I have configured my `Startup.cs` like:

#### Startup.cs
```C#
        public void ConfigureServices(IServiceCollection services)
        {
            // ...
            services.AddScoped<UserInfoAccessor, UserInfoAccessor>();
            services.AddScoped<UserInfo, UserInfo>();
            services.AddMvc();
            services.AddHttpContextAccessor();

            services.AddEntityFrameworkSqlServer();
            services.AddDbContextPool<MyContext>((serviceProvider, options) =>
            {
                options.UseSqlServer(connectionString);
                options.UseInternalServiceProvider(serviceProvider);
            });
        }

        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            // ...
            app.UseUserInfoMiddleware();
            app.UseMvc();
        }
```
#### UserInfoMiddleware.cs
```C#
    public class UserInfoMiddleware
    {

        private readonly RequestDelegate _next;

        public UserInfoMiddleware(RequestDelegate next)
        {
            _next = next;
        }

        public async Task InvokeAsync(HttpContext context, UserInfoAccessor userInfoAccessor)
        {
            // Here I will have my logic to fetch user data and set it on the middleware
            userInfoAccessor.UserInfo.Test = $"MODIFIED BY MIDDLEWARE - ${context.Request.QueyString}";
            await _next(context);
        }

    }
```
#### UserInfoAccessor.cs
```C#
    public class UserInfoAccessor
    {
        public UserInfo UserInfo { get; set; }

        public DateTime Created { get; }

        public UserInfoAccessor(UserInfo userInfo)
        {
            Created = DateTime.Now;
            UserInfo = userInfo;
        }
    }
```
#### UserInfo.cs
```C#
    public class UserInfo
    {
        public string Test { get; set; }
    }
```
#### TestController.cs
```C#
    public class TestController : Controller
    {
        private UserInfoAccessor _userInfoAccessor;

        public TestController(UserInfoAccessor userInfoAcessor)
        {
            _userInfoAccessor = userInfoAcessor;
        }

        [HttpGet("test")]
        public void Test()
        {
            _userInfoAccessor.UserInfo.Test = $"MODIFIED BY CONTROLLER - {HttpContext.Request.QueryString}";
            _context.TestService();
        }
    }
```
#### MyContext.cs
```C#
    public class MyContext : DbContext
    {       
        // ...
        public void TestService()
        {
            var userInfo1= this.GetService<UserInfoAccessor>(); 
            var userInfo2 = this.GetInfrastructure().GetService<UserInfoAccessor>();
            var queryString = this.GetService<IHttpContextAccessor>().HttpContext.Request.QueryString;
            var i = 0;
        }
     }
```

### Expected
Calling the `TestService()` from a `TestController` I can see that `queryString` assumes different values according to my request. That's what I expect for my `UserInfo`, since it is defined as `Scoped`, but it get's the same value for every request (same `DateTime` and `Test` set to `null`).

Basically, at the `Middleware` it changes the `Test` value. On the `Controller` I get the same instance as I do on the `Middleware`, so it just updates the `Test` value. But then on the the `MyContext.cs` I get the same instance of `UserInfo` everytime, with `Test` set to null and an old `DateTime` (notice I'm using DbPool).

Why I get the same instance everytime if my `UserInfoAccessor` is added as `Scoped`?

### Further technical details
EF Core version: 2.2.2
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10 Pro
IDE: Visual Studio 2017

**Comment by ajcvickers** [comment]
Duplicate of #14169

**Comment by Menighin** [comment]
@ajcvickers thanks. I understand I am not able to use DbPooling in this scenario. 
Just a question: Do you know, from a design point of view, why the `HttpContextAccessor` works fine? I dig a bit but couldn't find why it works... I saw it is registered as a `Singleton` and it has a private static proprerty that I'm guessing is keeping a dictionary of all the requests and its `HtppContext`s...?

**Comment by ajcvickers** [comment]
@Menighin if `IHttpContextAccessor` is registered as a singleton, then there shouldn't be any problem because you'll get the same instance resolved from any scope, so it doesn't matter which scope the `DbContext` was originally resolved from.

**Comment by Menighin** [comment]
@ajcvickers that's not what I meant, sorry haha.
When I use `services.AddHttpContextAccessor()` (native from .Net) I can see in the source code that it is adding it as a `Singleton`. But somehow .NET resolves the dependency right when I use `this.GetService<IHttpContextAccessor>().HttpContext`. I get the `HttpContext` scoped for that request (you can see in my code that it gets the correct `QueryString` for each request).
I was asking if you know how this is done behind the scenes because maybe I could use the same approach in my scenario.

**Comment by ajcvickers** [comment]
@Menighin I'm assuming it's because there is only ever one request per thread, and hence the current HttpContext can be stored in and retrieved from an AsyncLocal. If it allows you to get the service provider for the current request, then it certainly seems like you could use this to get your service. But you might be better off asking an ASP.NET Core expert. :-)

**Comment by Menighin** [comment]
Thanks @ajcvickers.
I did using `AsyncLocal` and it seems to be working just fine.
For anyone interested, basically I changed my `UserInfoAcessor` to:

#### UserInfoAccessor.cs
    public class UserInfoAccessor : IUserInfoAccessor
    {
        private static readonly AsyncLocal<ValueTuple<string, UserInfo>> _userInfo = new AsyncLocal<ValueTuple<string, UserInfo>>();

        public UserInfo UserInfo
        {
            get
            {
                var (item1, userInfo) = _userInfo.Value;
                return item1 != userInfo?.TraceIdentifier ? null : userInfo;
            }
            set => _userInfo.Value = new ValueTuple<string, UserInfo>(value?.TraceIdentifier, value);
        }
    }

And in my Middleware, I create a new instance based on the `TraceIdentifier` from the `HttpContext`:

#### UserInfoMiddleware.cs
        public async Task InvokeAsync(HttpContext context, GcoContext ctx, IUserInfoAccessor userInfoAccessor)
        {
            userInfoAccessor.UserInfo = new UserInfo
                {Test = "Modified by Middleware", TraceIdentifier = context.TraceIdentifier};
            await _next(context);
        }

And I injected my `UserInfoAccessor` as a `Singleton`.

This seems to be working fine, have done some tests on multiple requests being fire on the same time and it seems to handle just fine.
If anyone sees any problem with it please let me know!
